enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx

```diff
@@ -0,0 +1,187 @@
+import { useDisclosure } from "@mantine/hooks";
+import { useMemo, useState } from "react";
+import { match } from "ts-pattern";
+
+import S from "embedding-sdk/components/private/InteractiveQuestion/components/Picker.module.css";
+import { MultiStepPopover } from "embedding-sdk/components/private/util/MultiStepPopover";
+import type { UpdateQueryHookProps } from "metabase/query_builder/hooks";
+import { getFilterItems } from "metabase/querying/filters/components/FilterPanel/utils";
+import { FilterPicker } from "metabase/querying/filters/components/FilterPicker";
+import type { FilterColumnPickerProps } from "metabase/querying/filters/components/FilterPicker/FilterColumnPicker";
+import * as Lib from "metabase-lib";
+
+import { useInteractiveQuestionContext } from "../../context";
+import { BadgeList } from "../util/BadgeList";
+import { ToolbarButton } from "../util/ToolbarButton";
+
+const useFilterHandlers = ({
+  query,
+  stageIndex,
+  onQueryChange,
+}: UpdateQueryHookProps) => {
+  const getFilterName = (filter: Lib.FilterClause) => {
+    return Lib.displayInfo(query, stageIndex, filter).longDisplayName;
+  };
+
+  const onAddFilter = (filter: Lib.Filterable) => {
+    const nextQuery = Lib.filter(query, stageIndex, filter);
+    onQueryChange(nextQuery);
+  };
+
+  const onRemoveFilter = (filterClause: Lib.FilterClause) => {
+    const nextQuery = Lib.removeClause(query, stageIndex, filterClause);
+    onQueryChange(nextQuery);
+  };
+
+  const onUpdateFilter = (
+    currentFilterClause: Lib.FilterClause,
+    nextFilterClause: Lib.Filterable,
+  ) => {
+    const nextQuery = Lib.replaceClause(
+      query,
+      stageIndex,
+      currentFilterClause,
+      nextFilterClause,
+    );
+    onQueryChange(nextQuery);
+  };
+
+  return {
+    getFilterName,
+    onAddFilter,
+    onRemoveFilter,
+    onUpdateFilter,
+  };
+};
+
+const FilterInner = ({
+  query,
+  onQueryChange,
+  stageIndex,
+  withColumnItemIcon,
+  withColumnGroupIcon,
+  withCustomExpression,
+}: UpdateQueryHookProps &
+  Pick<
+    FilterColumnPickerProps,
+    "withColumnItemIcon" | "withColumnGroupIcon" | "withCustomExpression"
+  >) => {
+  const { getFilterName, onAddFilter, onRemoveFilter, onUpdateFilter } =
+    useFilterHandlers({ query, onQueryChange, stageIndex });
+
+  const filters = useMemo(() => getFilterItems(query), [query]);
+
+  const [step, setStep] = useState<"list" | "picker">("list");
+
+  const [opened, { close, toggle }] = useDisclosure(false, {
+    onOpen: () => {
+      setStep(filters.length === 0 ? "picker" : "list");
+    },
+  });
```

Comment:
Just an observation that I saw this block across multiple components while reviewing this stack of PRs, I'm thinking if we can generalize the notion of "dropdown" across these components.

Reply:
It's definitely something I want to do in the future once all of this is done - I want to make it work before it has to work well 🙂